Better support for Omega in Polaris

xylar opened the pull request:
<!--
Thank you for your pull request.
Please add a description of what is accomplished in the PR here at the top:
-->

Most of the work in this PR is focused on support for mapping from MPAS-Ocean namelist options (in YAML format) to their Omega equivalents.  This support has required changes to the `polaris.yaml` and `polaris.model_step` framework as well as extensive changes to the Ocean framework and tasks.

The proposed solution would allow for:
1. shared model config options (using MPAS-Ocean names) in an `ocean:` section in a yaml file or with `config_model='ocean'` when added in code.
2. model config options specific to MPAS-Ocean (using MPAS-Ocean names and ignored by Omega) with an `mpas-ocean:` section in a yaml file or with `config_model='mpas-ocean'` when added in code.
2. model config options specific to Omega (using Omega names and ignored by MPAS-Ocean) with an `omega:` section in a yaml file or with `config_model='omega'` when added in code.

Here is an example:
```yaml
ocean:
  time_management:
    config_run_duration: {{ run_duration }}
  time_integration:
    config_dt: {{ dt }}
    config_time_integrator: {{ time_integrator }}

mpas-ocean:
  bottom_drag:
    config_bottom_drag_mode: implicit
    config_implicit_bottom_drag_type: constant
    config_implicit_constant_bottom_drag_coeff: 0.0
  manufactured_solution:
     config_use_manufactured_solution: true
  debug:
    config_disable_vel_hmix: true

omega:
  Tendencies:
    VelDiffTendencyEnable: false
    VelHyperDiffTendencyEnable: false
```

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.md`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://e3sm-project.github.io/polaris/main/developers_guide/building_docs.html) and changes look as expected
* [x] `Testing` comment in the PR documents testing used to verify the changes

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

xylar commented:
@cbegeman, @hyungyukang and @sbrus89, I'm still testing and need to build the docs locally to make sure they're okay but I think this is working for MPAS-Ocean still (I can set up all ocean tests and those I've run so far all work) and I'm also able to set up the `manufactured_solution` test with `--model=omega` as long as I manually put an `omega.yml` file next to the `omega.exe`.

Let me know what you all think.  Whatever testing you'd like to do would be appreciated but also just a look at (parts of) the code would be fine.  I realize a lot changed so maybe the more important thing might be to look at `manufactured_solution` and treat it as an example of what these changes would mean for other tests.

xylar commented:
## Testing

I set up the `manufactured_solution` test with `--model=omega` after building Omega according to the instructions I added to the developer's guide in this PR and manually adding an `omega.yml` next to the `omega.exe` in:
```
e3sm_submodules/Omega/components/omega/build/src
```
The resulting `omega.yml` files looked as I expected, e.g.:
```
omega:
  TimeManagement:
    StartTime: 0001-01-01_00:00:00
    StopTime: 0001-01-01_10:00:00
    RunDuration: 0000_10:00:00.000
    CalendarType: No Leap
  TimeIntegration:
    TimeStepper: RK4
    TimeStep: 0000_00:10:00.000
  Decomp:
    HaloWidth: 3
    DecompMethod: MetisKWay
  State:
    NTimeLevels: 2
  Advection:
    FluxThicknessType: Center
  Tendencies:
    ThicknessFluxTendencyEnable: true
    PVTendencyEnable: true
    KETendencyEnable: true
    SSHTendencyEnable: true
    VelDiffTendencyEnable: false
    ViscDel2: 1.0e3
    VelHyperDiffTendencyEnable: false
    ViscDel4: 1.2e11
```

I also ran all the ocean test cases (yes, all 107) on Chrysalis with MPAS-Ocean.  I am seeing some failures in the 1 km isomip_plus test cases that are related to remapping and seemingly not related to this PR.  All other tests passed.  See:
```
/lcrc/group/e3sm/ac.xylar/polaris_0.4/chrysalis/test_20240904/all_omega_map
```

xylar commented:
Okay, I'm also done with testing and docs so this is fully ready for review.

hyungyukang commented:
I was able to test this branch by combining the Omega Driver PR (https://github.com/E3SM-Project/Omega/pull/123).

In the driver PR, I had to make some changes to obtain elapsed time (from 0 sec) during time integration and fix the three digit for the second in the time format.

I tried to test this PR and Omega with GNU compiler on pm-cpu, but It doesn't work. It seems there is a problem with the time manager, but I need to investigate more.

I moved on to NVIDIA compiler for both Omega and Polaris on pm-cpu, and it was successful. The dropdown below is to show processes to retrieve the test I've done on pm-cpu. Note that I added a manufactured solution group in `omega/config/DriverTest.yml`: 
```
ManufacturedSolution:
     UseManufacturedSolution: true
```

<details>
<summary>Process to retrieve the test</summary>

Download Omega-Polaris test codes on Perlmutter
```
git clone git@github.com:hyungyukang/Omega.git
cd Omega
git checkout hkang/omega/driver-Polaris-test
git submodule update --init --recursive externals/ekat externals/scorpio cime
````

Install Omega with NVIDIA compiler on pm-cpu
```
cd component/omega
mkdir build
cd build
module load cmake

./omega_build_nvidia_pm-cpu.sh
./omega_build.sh
```

Copy YAML config file
```
cp ../config/DriverTest.yml ./src/omega.yml
```

Install Polaris (add-omega-config-map PR) with NVIDIA compiler on Perlmutter
```
git clone git@github.com:E3SM-Project/polaris.git
cd polaris
git checkout add-omega-config-map
git submodule update --init --recursive

./configure_polaris_envs.py --conda /global/cfs/projectdirs/e3sm/hgkang/miniconda3 -c nvidia -m pm-cpu
```
Polaris setup for Omega model with the manufactured solution
```
cd polaris
source load_dev_polaris_0.4.0-alpha.2_pm-cpu_nvidia_mpich.sh
polaris setup -t ocean/planar/manufactured_solution -m pm-cpu -w /global/cfs/projectdirs/e3sm/hgkang/polaris/work25 -p  /pscratch/sd/h/hgkang/Omega_test_240908/components/omega/build --model omega
```
Submit a job
```
cd /global/cfs/projectdirs/e3sm/hgkang/polaris/work25
sbatch job_script.custom.sh
```
</details>

After the job was completed, I was able to see `PASS` in the log file and the second-order convergence rate as shown in the figure below.
```
ocean/planar/manufactured_solution
  * step: init_200km
          execution:        SUCCESS
          runtime:          0:00:08
  * step: forward_200km
          execution:        SUCCESS
          runtime:          0:00:09
  * step: init_100km
          execution:        SUCCESS
          runtime:          0:00:03
  * step: forward_100km
          execution:        SUCCESS
          runtime:          0:00:08
  * step: init_50km
          execution:        SUCCESS
          runtime:          0:00:04
  * step: forward_50km
          execution:        SUCCESS
          runtime:          0:00:18
  * step: init_25km
          execution:        SUCCESS
          runtime:          0:00:08
  * step: forward_25km
          execution:        SUCCESS
          runtime:          0:02:01
  * step: analysis
          execution:        SUCCESS
          runtime:          0:00:02
  task execution:   SUCCESS
  task runtime:     0:03:00
Task Runtimes:
0:03:00 PASS ocean/planar/manufactured_solution
Total runtime: 0:03:32
PASS: All passed successfully!
```

<img width="500" alt="image" src="https://github.com/user-attachments/assets/d7dc733d-c2c1-494c-bee8-b71c995b5c65">

Based on my test, this branch is working properly. I had to copy `omega.yml` manually from `config/DriverTest.yml`, but I believe that this process can be fixed straightforwardly or changed easily.

cbegeman commented on polaris/ocean/model/ocean_model_step.py:
```diff
@@ -214,3 +258,148 @@ def _update_ntasks(self):
+                    break
+
+            if option_found:
+                out_section = omega_seciton
```
```suggestion
                out_section = omega_section
```

cbegeman commented on polaris/ocean/model/ocean_model_step.py:
```diff
@@ -214,3 +258,148 @@ def _update_ntasks(self):
+        for entry in self.map:
+            section_found = False
+            section_dict = entry['section']
+            for mpaso_section, omega_seciton in section_dict.items():
```
```suggestion
            for mpaso_section, omega_section in section_dict.items():
```

cbegeman reviewed (approved):
@xylar I have a few minor comments/suggestions. I'm going out of town Thurs morning so feel free to merge whenever you're happy with the reviews you've gotten. I'm approving based on visual inspection and the tests you've run.

xylar commented on polaris/ocean/tasks/baroclinic_channel/forward.py:
```diff
@@ -86,7 +86,8 @@ def __init__(self, component, resolution, name='forward', subdir=None,
             # forward.yaml
-            self.add_model_config_options(options=dict(config_mom_del2=nu))
+            self.add_model_config_options(options=dict(config_mom_del2=nu),
+                                          config_model='mpas-ocean')
```
I think we want to keep everything outside of `manufactured_solution` as strictly `mpas-ocean` for now.  We can revisit as we verify support for Omega in each test.

xylar commented on polaris/ocean/tasks/internal_wave/forward.py:
```diff
@@ -76,12 +76,14 @@ def __init__(self, component, name='forward', subdir=None, indir=None,
             # forward.yaml
-            self.add_model_config_options(options=dict(config_mom_del2=nu))
+            self.add_model_config_options(options=dict(config_mom_del2=nu),
+                                          config_model='mpas-ocean')
```
Again, I think we want to explicitly evaluate Omega support in a separate PR and assume only MPAS-Ocean support for now.

xylar commented on polaris/ocean/tasks/internal_wave/forward.py:
```diff
@@ -136,4 +138,5 @@ def dynamic_model_config(self, at_setup):
                 time.strftime('%H:%M:%S', time.gmtime(run_seconds))
-        self.add_model_config_options(options=options)
+        self.add_model_config_options(options=options,
+                                      config_model='mpas-ocean')
```
Same as above.

xylar commented on polaris/ocean/model/ocean_model_step.py:
```diff
@@ -214,3 +258,148 @@ def _update_ntasks(self):
+                break
+
+        if not option_found:
+            raise ValueError(f'No mapping found for {section}/{option}')
```
I don't see any reason not to prefer your suggestion over my code.  I'll make that change.

xylar replied:
Ah, I see.  The issue is that the map is a list of dictionaries.  We have to iterate through the list (the outer loop) and then loop over options if this list entry has the right MPAS-Ocean section name.  So it's not a simple dictionary in the way you're thinking it might be.

The reason that it's not a simple dictionary is that it's a mapping of both sections and options from MPAS-Ocean to Omega.  I didn't see a good way to do this as a simple dictionary.  There wasn't a nice way to list both the section and option for both the inputs and outputs to the dictionary, especially because of the redundancy.  At the same time, there are some MPAS-Ocean config options that are spread across multiple Omega config sections and visa versa.

So I think we're stuck with this relatively clunky approach unless you have a different data structure for the map itself that you would recommend.

cbegeman replied:
See more thorough code changes in other suggestion

xylar commented:
We need to update the Omega submodule here once https://github.com/E3SM-Project/Omega/pull/123/ gets merged.

cbegeman commented on polaris/ocean/model/ocean_model_step.py:
```diff
@@ -214,3 +258,148 @@ def _update_ntasks(self):
+            if option_found:
+                out_section = omega_seciton
+                out_option = omega_option
+                break
```
I tested this version for the current config options for omega and a config option that doesn't map and it seems to perform fine
```suggestion
        option_found = False

        # traverse the map
        for entry in self.map:
            section_dict = entry['section']
            try:
                omega_section = section_dict[section]
            except KeyError:
                continue
            else:
                options_dict = entry['options']
                option_found = False
                try:
                    omega_option = options_dict[option]
                except KeyError:
                    continue
                else:
                    option_found = True
                    out_section = omega_section
                    out_option = omega_option
                    break
```

xylar replied:
Okay, I made this suggested change.  Thanks @cbegeman.  I'll test it again once https://github.com/E3SM-Project/Omega/pull/123 has been merged and the Omega submodule has been update.

mark-petersen commented:
I was able to run the manufactured solution test case successfully on Frontier with Omega.

<details>
<summary> Here is my list of commands </summary> <p>

```

### polaris instructions ###
cd /ccs/home/mpetersen/repos/polaris/pr
./configure_polaris_envs.py   --conda ${HOMEDIR}/miniforge3 --compiler crayclang
source load_dev_polaris_0.4.0-alpha.2_frontier_crayclang_mpich.sh

git submodule update --init e3sm_submodules/Omega
cd e3sm_submodules/Omega/
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime

git submodule update --init e3sm_submodules/E3SM
cd components/mpas-ocean
git submodule update --init --recursive src/cvmix/ src/gotm/ src/BGC src/MARBL src/gotm src/FFTW src/SHTNS src/ppr

polaris list |grep man
  56: ocean/planar/manufactured_solution


polaris setup -p /lustre/orion/cli115/scratch/mpetersen/runs/240910_omega/build -w $r/240911_omega_polaris_test --model omega -n 56

salloc -A cli115 -J inter -t 2:00:00 -q debug -N 1 -S 0
cd /ccs/home/mpetersen/repos/polaris/pr
source load_dev_polaris_0.4.0-alpha.2_frontier_crayclang_mpich.sh
cd /lustre/orion/cli115/scratch/mpetersen/runs/240911_omega_polaris_test
cd ocean/planar/manufactured_solution/
polaris serial

```
</details>

My main command was
```
polaris setup \
  -p /lustre/orion/cli115/scratch/mpetersen/runs/240910_omega/build \
  -w /lustre/orion/cli115/scratch/mpetersen/runs/240911_omega_polaris_test \
  --model omega \
  -n 56
```
Note that I had already built omega for frontier cpu with these commands

<details>
<summary> Frontier cpu build commands</summary> <p>

```
CODE_DIR=opr
RUNDIR=240910_omega_gpu
mkdir /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR
cd !$

cd /ccs/home/mpetersen/repos/E3SM/${CODE_DIR}
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd /lustre/orion/cli115/scratch/mpetersen/runs/$RUNDIR

module load cmake
rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=/ccs/proj/cli115/software/polaris/frontier/spack/dev_polaris_0_4_0_crayclanggpu_mpich/var/spack/environments/dev_polaris_0_4_0_crayclanggpu_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=crayomegagpu \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /ccs/home/mpetersen/repos/E3SM/${CODE_DIR}/components/omega -B .
./omega_build.sh
```
</details>

I had to do two additional things to get the tests to run:
1. The Omega code looks for the input file `OmegaMesh.nc`, but the file `init.nc` was provided by Polaris. So I made a quick link in each forward directory like this:
```
cd /lustre/orion/cli115/scratch/mpetersen/runs/240911_omega_polaris_test/
cd ocean/planar/manufactured_solution/forward/200km
ln -isf ../../init/200km/initial_state.nc OmegaMesh.nc
```
2. The file OmegaMesh.nc has 1 vertical layer and the default in `omega.yml` is `NVertLevels: 60`. Omega produces a pio error, but that is fixed if I set `NVertLevels: 1` in the local `omega.yml`.

I think both of these will be corrected once we have https://github.com/E3SM-Project/Omega/pull/123 merged and better control over file names with io-streams. With those two fixes I was able to go into each directory and run the simulation, like this:
```
salloc -A cli115 -J inter -t 2:00:00 -q debug -N 1 -S 0
cd /ccs/home/mpetersen/repos/polaris/pr
source load_dev_polaris_0.4.0-alpha.2_frontier_crayclang_mpich.sh
cd /lustre/orion/cli115/scratch/mpetersen/runs/240911_omega_polaris_test
cd ocean/planar/manufactured_solution/init/200km/
polaris serial
cd ../../forward/200km
polaris serial
```
and it runs successfully. It does not yet have an `output.nc` file to run analysis on, but that will come once i/o streams are in.

mark-petersen reviewed (approved):
Approving based on my testing above as well as @xylar and @hyungyukang's tests. 

Thanks @xylar for your fast work here. I think we will continue to update Polaris in future PRs as the Omega i/o streams are finalized.

mark-petersen commented:
@xylar, in my testing above I did this sequence to build Omega myself: 
```
export PARMETIS_ROOT=/ccs/proj/cli115/software/polaris/frontier/spack/dev_polaris_0_4_0_crayclanggpu_mpich/var/spack/environments/dev_polaris_0_4_0_crayclanggpu_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=crayomegagpu \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /ccs/home/mpetersen/repos/E3SM/${CODE_DIR}/components/omega -B .
./omega_build.sh
```
and then I point to that directory with the `polaris setup -p` flag. Is that the intended workflow, or is there a polaris command that builds the code within the submodule directories?

hyungyukang reviewed (approved):
I'm approving this PR based on my testing, @mark-petersen and @xylar 's tests, and @cbegeman 's visual inspection. During my testing with Omega-0 connected to Polaris, I didn't have to do much on the Polaris side. It was straightforward.

Thanks @xylar for your efforts!

xylar commented:
I have had to make some more changes here following https://github.com/E3SM-Project/Omega/pull/123:
* I have switched the `config_model` from `omega` to `Omega`.  This is the model name used in YAML files and associated calls to change YAML config options.  But I am keeping the model name elsewhere in Polaris as `omega`, e.g. `polaris setup --model=omega`
* I have had to set `config_stop_time = none` in a bunch of places.  This is the MPAS-Ocean default but not the Omega default, and we want to set config options that affect test cases in any case (rather than relying on the defaults).
* In the user's guide, I have added a short section about which model(s) each set of ocean tasks is compatible with (and also added this to the template).
* I have updated the Omega submodule
* I added a `configure()` method to manufactured solution to handle the different default time integrators for the 2 models.

I am currently on vacation so I haven't had a chance to do thorough testing.

@hyungyukang, could you re-test Omega with manufactured solutions?

@cbegeman, could you run the nightly suite and/or whatever tests you have the time and energy for?  I think for now it's sufficient that MPAS-Ocean tests aren't broken but you can certainly test if they're BFB, too, if you have the time and energy.

xylar commented:
@hyungyukang, when I try to run `omega.exe` (using Polaris), I'm seeing:
```
PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: Start+count exceeds dimension bound (file = OmegaMesh.nc) (/home/ac.xylar/e3sm_work/omega/omega/fix-run-duration-stop-time-inconsistent/externals/scorpio/src/clib/pio_darray_int.c: 1697)
Obtained 10 stack frames.
/lcrc/group/e3sm/ac.xylar/polaris_0.4/chrysalis/test_20240916/omega_manufac_soln_fix_warning/ocean/planar/manufactured_solution/forward/200km/./omega.exe() [0x6b644d]
```
Did you see this when you tried to run?  If so, how did you get around it?  Is it an issue with the NetCDF format we're writing out?

hyungyukang commented:
> @hyungyukang, when I try to run `omega.exe` (using Polaris), I'm seeing:
> 
> ```
> PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: Start+count exceeds dimension bound (file = OmegaMesh.nc) (/home/ac.xylar/e3sm_work/omega/omega/fix-run-duration-stop-time-inconsistent/externals/scorpio/src/clib/pio_darray_int.c: 1697)
> Obtained 10 stack frames.
> /lcrc/group/e3sm/ac.xylar/polaris_0.4/chrysalis/test_20240916/omega_manufac_soln_fix_warning/ocean/planar/manufactured_solution/forward/200km/./omega.exe() [0x6b644d]
> ```
> 
> Did you see this when you tried to run? If so, how did you get around it? Is it an issue with the NetCDF format we're writing out?

@xylar , the error is probably caused by `NVertLevels` in `omega.yml`. During my testing, I had to change `NVertLevels=1` in `omega.yml` before copying it to Polaris. Also in `omega/src/base/Decomp.h` (line 237), I changed the initial file name from `OmegaMesh.nc` to `init.nc` as Polaris creates `init.nc` file as the initial condition.

hyungyukang commented:
I'm going to test this updated PR and the merged driver PR in Omega-0 (https://github.com/E3SM-Project/Omega/pull/123) at the same time by implementing the manufactured solution.

mark-petersen commented:
I also had to do the following, [as noted last week](https://github.com/E3SM-Project/polaris/pull/221#issuecomment-2344552473):
> The file OmegaMesh.nc has 1 vertical layer and the default in omega.yml is NVertLevels: 60. Omega produces a pio error, but that is fixed if I set NVertLevels: 1 in the local omega.yml.

Otherwise I get that error
```
PIO: FATAL ERROR: Aborting... FATAL ERROR: NetCDF: Start+count exceeds dimension bound (file = OmegaMesh.nc) (/home/ac.xylar/e3sm_work/omega/omega/fix-run-duration-stop-time-inconsistent/externals/scorpio/src/clib/pio_darray_int.c: 1697)
```

xylar commented:
@mark-petersen and @hyungyukang, the fix seems pretty trivial (though I didn't have time to test it).  Let me know if my last commit has fixed the issue.

xylar commented:
This needs a fix requested in https://github.com/E3SM-Project/Omega/issues/126.  Then, the `omega.cfg` file will need to be updated.

xylar commented on polaris/ocean/omega.cfg:
```diff
@@ -0,0 +1,24 @@
+# has been built
+component_path = ${paths:source_path}/build
+
+# The model_config section defines paths to yaml file with default model config
+# options that will be used to generate specific yaml config files for Omega.
+[model_config]
+defaults = ${paths:source_path}/configs/Default.yml
```
This didn't work as desired.  The `-p` flag only changes `component_path` so `source_path` stays pointing to the default location.  This will need a fix requested in https://github.com/E3SM-Project/Omega/issues/126 and then to be updated.

xylar replied:
```suggestion
# The model_config section defines paths to yaml file with default model config
# options that will be used to generate specific yaml config files for Omega.
[model_config]
defaults = ${paths:component_path}/configs/Default.yml
```

xylar commented on polaris/ocean/omega.cfg:
```diff
@@ -0,0 +1,24 @@
+[paths]
+
+
+# the relative or absolute path to the Omega component source code
+source_path = ${paths:polaris_branch}/e3sm_submodules/Omega/components/omega
+
+# the relative or absolute path to the root of a branch where Omega
+# has been built
+component_path = ${paths:source_path}/build
```
```suggestion
# the relative or absolute path to the root of a branch where Omega
# has been built
component_path = ${paths:polaris_branch}/e3sm_submodules/Omega/components/omega/build
```

mark-petersen commented:
After a conversation with @cbegeman  and @brian-oneill we would like to always read nVertLevels from the initial condition file. It takes a little more work up front to generate initial conditions, but once the workflow is set up the transition to Omega-1 will be seamless. We thought an optional YAML flag for nVertLevels was prone to user error.

I think the simplest way forward is to leave [e23ef33](https://github.com/E3SM-Project/polaris/pull/221/commits/e23ef330efa8b6119703693f13829911eb211df7) in so this works with the current head of `develop`. Once we have io streams working, we will have to revise the Omega driver and then Polaris again anyway. 

We can make the requested change in https://github.com/E3SM-Project/Omega/issues/126 and the `component_path` change above, and then I think this PR is ready to merge.

hyungyukang commented:
@xylar , I'm testing the latest version of this PR (commit=e23ef330). While testing, I found that Polaris does not change `TimeStepper` to `RK4` and `TimeStep` is too large for a given resolution. I didn't have this problem during the test last time (commit=.
I'm attaching two YAML files - `Default.yml` in `omega/config` and `omega.yml` in a Polaris run directory

Here is `Default.yml` in `omega/config/` but `ManufacturedSolution:` is added for testing the purpose.

```
Omega:
   TimeManagement:
      StartTime: 0001-01-01_00:00:00
      StopTime: 0001-01-01_02:00:00
      RunDuration: none #0000_02:00:00.000
      CalendarType: No Leap
   TimeIntegration:
      TimeStepper: Forward-Backward
      TimeStep: 0000_00:10:00
   Dimension:
      NVertLevels: 60
   Decomp:
      HaloWidth: 3
      DecompMethod: MetisKWay
   State:
      NTimeLevels: 2
   Advection:
      FluxThicknessType: Center
   Tendencies:
      ThicknessFluxTendencyEnable: true
      PVTendencyEnable: true
      KETendencyEnable: true
      SSHTendencyEnable: true
      VelDiffTendencyEnable: true
      ViscDel2: 1.0e3
      VelHyperDiffTendencyEnable: true
      ViscDel4: 1.2e11
   ManufacturedSolution:
      UseManufacturedSolution: true
```

And here is `omega.yml` in `work4/ocean/planar/manufactured_solution/forward/200km/` after setting up Polaris tests
```
Omega:
  TimeManagement:
    StartTime: 0001-01-01_00:00:00
    StopTime: none
    RunDuration: 0000_10:00:00.000 #0000_02:00:00.000
    CalendarType: No Leap
  TimeIntegration:
    TimeStepper: Forward-Backward
    TimeStep: 0000_01:40:00.000
  Dimension:
    NVertLevels: 1
  Decomp:
    HaloWidth: 3
    DecompMethod: MetisKWay
  State:
    NTimeLevels: 2
  Advection:
    FluxThicknessType: Center
  Tendencies:
    ThicknessFluxTendencyEnable: true
    PVTendencyEnable: true
    KETendencyEnable: true
    SSHTendencyEnable: true
    VelDiffTendencyEnable: false
    ViscDel2: 1.0e3
    VelHyperDiffTendencyEnable: false
    ViscDel4: 1.2e11
  ManufacturedSolution:
    UseManufacturedSolution: true
```

`RunDuration` and `NVertLevels` were changed correctly. In this test, `TimeStepper` should be `RK4` but was not changed. And `TimeStep` should be `0000_00:10:00.000`=600 seconds for 200 km but was set to `0000_01:40:00.000`=6000 seconds.

Do you have any idea on this or any suggestion? The instruction below is to retrieve the test I'm conducting. Please correct me if I've done anything wrong.

<details>
<summary>Instruction to retrieve the test</summary>
<br>

```
# Install Polaris on Perlmutter with NVIDIA compiler

git clone git@github.com:xylar/polaris.git
cd polaris
git checkout add-omega-config-map
git submodule update --init --recursive

# Update Omega-0 to test with Polaris
cd e3sm_submodules/Omega
git remote add test git@github.com:hyungyukang/Omega.git
git fetch test
git checkout hkang/omega/driver-merge-test

# Build Omega-0
cd component/omega
mkdir build
cd build

export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_4_0_nvidia_mpich/var/spack/environments/dev_polaris_0_4_0_nvidia_mpich/.spack-env/view

module load cmake

cmake \
   -DOMEGA_CIME_COMPILER=nvidia \
   -DOMEGA_BUILD_TYPE=Release \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -Wno-dev \
   -S ../ -B .

# Polaris setup with the manufactured solution
cd ../../../../../
source load_dev_polaris_0.4.0-alpha.2_pm-cpu_nvidia_mpich.sh
polaris setup -t ocean/planar/manufactured_solution -m pm-cpu -w /global/cfs/projectdirs/e3sm/hgkang/polaris/work4 --model omega
```

</details>

xylar commented:
@mark-petersen, I'm afraid we can't rush this.  Let's keep working at it until we get the kinks ironed out.

xylar commented:
@hyungyukang, I see.  I honestly didn't realize we had implemented RK4.  I  will remove the commit that explicitly set the time integrator for Omega to Forward-Backward.

xylar commented:
@hyungyukang, could you cherry-pick https://github.com/E3SM-Project/Omega/pull/128 onto whatever Omega branch you're using and try again with this Polaris branch?  I think it should be back to using RK4.  Sorry about that!

cbegeman commented:
For what it's worth, I ran the nightly MPAS-O test suite and all tests passed.

hyungyukang commented:
There is an inconsistency in the directory name for configuration files within the build directory between this Polaris pull request (using "build/configs") and Omega (using "build/config"). I addressed this issue in my comment on the E3SM/Omega pull request (E3SM-Project/Omega#128) and suggested modifying it on the Omega side, as the original directory name for configuration files is `configs`.

hyungyukang commented:
Although there is an inconsistency in the directory name for the configuration files as I mentioned above, I was able to re-test everything by manually matching the directory name. Fixing this issue is very straightforward.

I used the most recent version of Omega-0 in the `develop` branch (commit=01128e24b) to test this Polaris PR. I implemented the manufactured solution in Omega-0, wrote a code for output streams, and added `ManufacturedSolution: \ UseManufacturedSolution: true`  to `configs/Default.yml` for the test purpose.

Here is an instruction to retrieve the test what I've done:
<details>
<summary>Instruction for the test done for this PR</summary>
<br>

```
#Install Polaris (this PR) on Perlmutter with NVIDIA compiler
git clone git@github.com:xylar/polaris.git
cd polaris
git checkout add-omega-config-map
git submodule update --init e3sm_submodules/Omega

# Obtain Omega-0 code modified for this test
cd e3sm_submodules/Omega
git remote add test git@github.com:hyungyukang/Omega.git
git fetch test
git checkout hkang/omega/driver-merge-test
git submodule update --init --recursive externals/ekat externals/scorpio cime

# Build Omega-0 with NVIDIA compiler
cd component/omega
mkdir build
cd build
export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_4_0_nvidia_mpich/var/spack/environments/dev_polaris_0_4_0_nvidia_mpich/.spack-env/view
module load cmake
cmake -DOMEGA_CIME_COMPILER=nvidia -DOMEGA_BUILD_TYPE=Release -DOMEGA_CIME_MACHINE=pm-cpu -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT} -Wno-dev -S ../ -B .
./omega_build.sh

# Polaris setup with the manufactured solution
cd ../../../../../
./configure_polaris_envs.py --conda /global/cfs/projectdirs/e3sm/hgkang/miniconda3 -c nvidia -m pm-cpu
source load_dev_polaris_0.4.0-alpha.2_pm-cpu_nvidia_mpich.sh
polaris setup -t ocean/planar/manufactured_solution -m pm-cpu -w /global/cfs/projectdirs/e3sm/hgkang/polaris/work51 --model omega

# Submit a Polaris job
cd /global/cfs/projectdirs/e3sm/hgkang/polaris/work51
sbatch job_script.custom.sh

```

</details>

After the job was completed, I saw `PASS` in the log file and obtained the second-order convergence rate figure visualized by Polaris.

<details>
<summary>PASS/FAIL logs</summary>
<br>

```
ocean/planar/manufactured_solution
  * step: init_200km
          execution:        SUCCESS
          runtime:          0:00:14
  * step: forward_200km
          execution:        SUCCESS
          runtime:          0:00:20
  * step: init_100km
          execution:        SUCCESS
          runtime:          0:00:05
  * step: forward_100km
          execution:        SUCCESS
          runtime:          0:00:05
  * step: init_50km
          execution:        SUCCESS
          runtime:          0:00:05
  * step: forward_50km
          execution:        SUCCESS
          runtime:          0:00:10
  * step: init_25km
          execution:        SUCCESS
          runtime:          0:00:12
  * step: forward_25km
          execution:        SUCCESS
          runtime:          0:00:14
  * step: analysis
          execution:        SUCCESS
          runtime:          0:00:04
  task execution:   SUCCESS
  task runtime:     0:01:29
Task Runtimes:
0:01:29 PASS ocean/planar/manufactured_solution
Total runtime: 0:02:03
PASS: All passed successfully!
```
</details>

<img width="500" alt="image" src="https://github.com/user-attachments/assets/1e1db599-e9e6-44d9-a710-6d303cb211c9">

Thanks @xylar very much for fixing codes very quickly!

xylar commented:
Okay, once https://github.com/E3SM-Project/Omega/pull/128 gets merged, I'll update the `Omega` submodule once again and this will be ready to merge.

hyungyukang reviewed (approved):
I am approving this PR based on my tests with Omega-0 using the manufactured solution, as well as @cbegeman 's MPAS-O nightly suite test.

As @xylar mentioned, the Omega submodule will need to be updated once https://github.com/E3SM-Project/Omega/pull/128 is merged into the main branch. 

Although the current branch is sufficient for testing Omega with the manufactured solution, additional work may be required once the output streams in Omega are completed.

Thank you very much, @xylar , for your hard work on this PR!

hyungyukang commented:
An additional request for the next PR is that we may no longer need to perform graph partitioning (i.e., linking `graph.info` and `gpmetis graph.info ...`) for Omega, as this process is now handled internally.

xylar commented:
@hyungyukang, good call.  That should be easy enough to address in a follow-up PR.  The extra files aren't an issue, they're just unneeded clutter.

cbegeman commented:
@hyungyukang and @xylar Good point. I'll open up a PR with that change shortly.

xylar commented:
Okay, I've updated the Omega submodule one more time.  I'll merge this once CI passes.

xylar commented:
@cbegeman, thanks for getting on the redundant metis file issue!
